zcash_client_backend/src/tor/http/exchange_rate.rs

```diff
@@ -0,0 +1,150 @@
+use futures_util::{join, FutureExt};
+use rand::{seq::IteratorRandom, thread_rng};
+use serde::{Deserialize, Deserializer};
+use tracing::{error, trace};
+
+use crate::tor::{Client, Error};
+
+impl Client {
+    /// Fetches the latest USD/ZEC exchange rate.
+    ///
+    /// We query the ticker price of the following trading pairs:
+    /// - `ZECUSDT` on Binance
+    /// - `ZEC-USD` on Coinbase
+    /// - `zecusd` on Gemini
+    ///
+    /// Returns an error if all three exchange queries fail.
+    pub async fn get_exchange_rate_usd(&self) -> Result<f64, Error> {
+        // "Never go to sea with two chronometers; take one or three."
+
+        let binance_ticker = self
+            .get_json::<BinanceTicker>(
+                "https://api.binance.com/api/v3/ticker/price?symbol=ZECUSDT"
+                    .parse()
+                    .unwrap(),
+            )
+            .map(|res| res.map(|resp| resp.into_body()));
+
+        let coinbase_ticker = self
+            .get_json::<CoinbaseTicker>(
+                "https://api.exchange.coinbase.com/products/ZEC-USD/ticker"
+                    .parse()
+                    .unwrap(),
+            )
+            .map(|res| res.map(|resp| resp.into_body()));
+
+        let gemini_ticker = self
+            .get_json::<GeminiTicker>("https://api.gemini.com/v2/ticker/zecusd".parse().unwrap())
+            .map(|res| res.map(|resp| resp.into_body()));
```

Comment:
I have now observed Gemini returning 451 Unavailable for Legal Reasons. I've only seen it once though; I'm guessing a specific Tor exit node was blocked, and on a subsequent query the circuit used a different exit.

Reply:
It's still the most reliable data source, so Gemini is now excluded from being ignored.